### PR TITLE
add EfficientIP Device Manager dynamic inventory

### PR DESCRIPTION
**Pull request by achauvinhameau**
##### SUMMARY
migrate EfficientIP inventory

##### COMPONENT NAME
community.general.efficientip

##### ADDITIONAL INFORMATION
Add EfficientIP SOLIDserver IPAM as an inventory using search and filter function in DeviceManager

### CODE REVIEW

**Comment by felixfontein** [comment]
Where was this plugin migrated from?

**Comment by Akasurde** [comment]
@felixfontein This is a new inventory plugin. I recommended creating a PR here instead of ansible/ansible repo.

**Comment by felixfontein** [comment]
I was just wondering whether it was migrated from another collection (and originally from ansible/ansible's pre-ansible-base), or from a PR.

**Comment by achauvinhameau** [comment]
It is brand new and was originally proposed in ansible repo this week-end.

Le mar. 14 avr. 2020 à 07:40, Felix Fontein <notifications@github.com> a
écrit :

File: plugins/inventory/efficientip.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+            type: dictionary
+        group_metadata:
+            description: regroup devices by metadata
+            required: False
+            type: list
+'''
+
+EXAMPLES = '''
+# example with only credentials to get the whole content of Device Manager
+plugin: efficientip
```
Comment (gundalow) [inline plugins/inventory/efficientip.py:80]: Do we need FQCN for new inventory plugin names?

File: plugins/inventory/efficientip.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+                                        filters=self.device_filter,
+                                        metadatas=self.group_metadata)
+
+        if cache_needs_update or (not cache and self.get_option('cache')):
+            self._cache[cache_key] = adevs
+
+        self._populate(adevs)
+
+#
+# ansible-test sanity --test pylint lib/ansible/plugins/inventory/efficientip.py
```
Comment (gundalow) [inline plugins/inventory/efficientip.py:358]: Nice to add notes about testing needed.
Paths need updating

File: plugins/inventory/efficientip.py (hunk 1 above)
Comment (felixfontein) [reply plugins/inventory/efficientip.py:80]: We do.

File: plugins/inventory/efficientip.py (hunk 2 above)
Comment (achauvinhameau) [reply plugins/inventory/efficientip.py]: updated

File: plugins/inventory/efficientip.py (hunk 1 above)
Comment (achauvinhameau) [reply plugins/inventory/efficientip.py]: done

File: plugins/inventory/efficientip.py [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+    name: efficientip
+    plugin_type: inventory
+    short_description: inventory based on EfficientIP IPAM
+    requirements:
+        - SOLIDserverRest > 2.0.1
+        - python 3
+    extends_documentation_fragment:
+        - inventory_cache
+    description:
+        - Get inventory hosts from EfficientIP IPAM and Device Manager
```
Comment (felixfontein) [inline plugins/inventory/efficientip.py:44]:
```suggestion
        - Get inventory hosts from EfficientIP IPAM and Device Manager.
```
Descriptions should always be sentences.

File: plugins/inventory/efficientip.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+    author:
+        - Alex Chauvin (@efficientip)
+    options:
+        plugin:
+            description: Token that ensures this is a source file for
+                         the plugin.
+            required: True
+            choices: ['efficientip']
+        api:
+            description: information to get connected to the SOLIDserver
```
Comment (felixfontein) [inline plugins/inventory/efficientip.py:55]:
```suggestion
            description: Information to get connected to the SOLIDserver.
```

File: plugins/inventory/efficientip.py [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+            required: True
+            type: dictionary
+        space:
+            description: name of the space
```
Comment (felixfontein) [inline plugins/inventory/efficientip.py:59]:
```suggestion
            description: Name of the space.
```

File: plugins/inventory/efficientip.py [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+            required: False
+            type: string
+            default: Local
+        filter:
+            description: list of filters to reduce the set of devices
```
Comment (felixfontein) [inline plugins/inventory/efficientip.py:64]:
```suggestion
            description: List of filters to reduce the set of devices
```

File: plugins/inventory/efficientip.py [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+                         returned from Device Manager. Can mix limit,
+                         device_in_subnet, device_of_class,
+                         device_metadata, device_intf_space. All
+                         filter are additiionnal, all conditions must
+                         match
+            required: False
+            type: dictionary
```
Comment (felixfontein) [inline plugins/inventory/efficientip.py:71]:
```suggestion
            type: dict
```

File: plugins/inventory/efficientip.py [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+            required: True
+            choices: ['efficientip']
+        api:
+            description: information to get connected to the SOLIDserver
+            required: True
+            type: dictionary
+        space:
+            description: name of the space
+            required: False
+            type: string
```
Comment (felixfontein) [inline plugins/inventory/efficientip.py:61]:
```suggestion
            type: str
```

**Comment by felixfontein** [comment]
(Please note there are more issues of the same kind, I didn't put comments everywhere :) )

File: plugins/inventory/efficientip.py [hunk 9]
```diff
@@ -0,0 +1,367 @@
+# -*- Mode: Python; python-indent-offset: 4 -*-
+#
+# Time-stamp: <2020-07-07 18:18:40 alex>
+#
+
+"""
+ EfficientIP inventory for ansible.
+ SOLIDserver Device Manager connection to gather devices filtered.
+"""
```
Comment (Andersson007) [inline plugins/inventory/efficientip.py:9]:
```suggestion
# -*- Mode: Python; python-indent-offset: 4 -*-
#
# Time-stamp: <2020-07-07 18:18:40 alex>
#

"""
 EfficientIP inventory for ansible.
 SOLIDserver Device Manager connection to gather devices filtered.
"""
```
could you please make it in a more standard way? Look into other plugins, e.g. in filter/time.py
```
  1 # -*- coding: utf-8 -*-
  2 # Copyright (c) 2020, René Moser <mail@renemoser.net>
  3 # GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
  4
  5 from __future__ import (absolute_import, division, print_function)
  6 __metaclass__ = type
```

File: plugins/inventory/efficientip.py [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+                         the plugin.
+            required: True
+            choices: ['efficientip']
+        api:
+            description: Information to get connected to the SOLIDserver.
+            required: True
+            type: dict
+        space:
+            description: Name of the space.
+            required: False
```
Comment (Andersson007) [inline plugins/inventory/efficientip.py:62]:
```suggestion
```
this is a default, so no needed

File: plugins/inventory/efficientip.py [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+            type: str
+            default: Local
+        filter:
+            description: List of filters to reduce the set of devices
+                         returned from Device Manager. Can mix limit,
+                         device_in_subnet, device_of_class,
+                         device_metadata, device_intf_space. All
+                         filter are additiionnal, all conditions must
+                         match.
+            required: False
```
Comment (Andersson007) [inline plugins/inventory/efficientip.py:72]:
```suggestion
```

File: plugins/inventory/efficientip.py [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+            type: dict
+        group_metadata:
+            description: Regroup devices by meta-data.
+            required: False
```
Comment (Andersson007) [inline plugins/inventory/efficientip.py:76]:
```suggestion
```

File: plugins/inventory/efficientip.py [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+            type: str
+            default: Local
+        filter:
+            description: List of filters to reduce the set of devices
+                         returned from Device Manager. Can mix limit,
+                         device_in_subnet, device_of_class,
+                         device_metadata, device_intf_space. All
+                         filter are additiionnal, all conditions must
+                         match.
```
Comment (Andersson007) [inline plugins/inventory/efficientip.py:71]:
```suggestion
            description:
            - List of filters to reduce the set of devices returned from Device Manager.
            - Can mix C(limit), C(device_in_subnet), C(device_of_class), C(device_metadata), C(device_intf_space).
            - All filter are additional, all conditions must match.
```

File: plugins/inventory/efficientip.py [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+# filter devices with interface in the specified space
+filter:
+ - device_intf_space: AWS
+
+# use metadata for grouping, the group key1 will contain
+# subgroup for each value (eg key1_on and key1_off)
+group_metadata:
+ - key1
+ - key2
+
```
Comment (Andersson007) [inline plugins/inventory/efficientip.py:121]:
```suggestion
```

File: plugins/inventory/efficientip.py [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+        if cache_needs_update or (not cache and self.get_option('cache')):
+            self._cache[cache_key] = adevs
+
+        self._populate(adevs)
+
+#
+# ansible-test sanity --test pylint plugins/inventory/efficientip.py
+# ansible-test sanity --test future-import-boilerplate plugins/inventory/efficientip.py
+# ansible-test sanity --test metaclass-boilerplate plugins/inventory/efficientip.py
```
Comment (Andersson007) [inline plugins/inventory/efficientip.py:367]:
```suggestion
```

**Review by Andersson007 — COMMENTED** [review]
@achauvinhameau hi, thanks for the PR!
If this is WIP, please mark it as `[WIP]`.
Would be also cool to have unit tests if possible.

**Comment by gundalow** [comment]
@achauvinhameau Anything we can do to help this progress

needs_info

**Comment by felixfontein** [comment]
Closing since there was no more reaction. @achauvinhameau if you're interested again in working on this, please ping me. If someone else is interested in continuing this PR, please also write it here so we can coordinate.

**Pull request closed without merging**